lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -235,105 +258,193 @@ def _restore_and_try_close_indices_if_needed(
         """Restores a backup to the current cluster."""
         if not self._can_unit_perform_backup(event):
             event.fail("Failed: backup service is not configured yet")
             return
 
+        if not self._check_if_restore_finished():
+            event.fail("Failed: previous restore is still in progress")
+            return
+        # Now, validate the backup is working
         backup_id = str(event.params.get("backup-id"))
+        if self._backup_is_available_for_restore(backup_id):
+            event.fail(f"Failed: no backup-id {backup_id}")
+            return
 
         # Restore will try to close indices if there is a matching name.
         # The goal is to leave the cluster in a running state, even if the restore fails.
         # In case of failure, then restore action must return a list of closed indices
         closed_idx = set()
         try:
-            backups = self._list_backups()
-            if backup_id not in backups.keys():
-                event.fail(f"Failed: no backup-id {backup_id}, options available: {backups}")
-                return
-            backup_id_state = self.get_snapshot_status(backups[backup_id]["state"])
-            if backup_id_state != BackupServiceState.SUCCESS:
-                event.fail(
-                    f"Failed: no backup-id {backup_id} not successful, state is: {backup_id_state}"
-                )
-                return
-            output, closed_idx = self._restore_and_try_close_indices_if_needed(backup_id)
+            closed_idx = self._close_indices_if_needed(backup_id)
+            output = self._issue_restore_request(
+                backup_id, event.params.get("wait-for-completion", False)
+            )
             logger.debug(f"Restore action: received response: {output}")
             logger.info(f"Restore action succeeded for backup_id {backup_id}")
-        except OpenSearchListBackupError as e:
-            event.fail(f"Failed: {e}, closed indices: {closed_idx}")
+        except OpenSearchRestoreFailedClosingIdxError as e:
+            event.fail(f"Failed: {e}")
             return
+
+        # Post execution checks
+        # Was the call successful?
         if self.get_service_status(output) != BackupServiceState.SUCCESS:
             status = str(self.get_service_status(output))
             event.fail(f"Failed with: {status}, closed indices: {closed_idx}")
+            return
+        # Finally, if status returns success, ensure all indices have been updated
+        # otherwise, raise an assert error and retry
+        if event.params.get("wait-for-completion", False):
+            # We are waiting for restore to complete, then we need to check the status
+            shards = output.get("snapshot", {}).get("shards", {})
+            assert shards.get("successful", -1) == shards.get("total", 0)
+        elif output.get("accepted", False):
+            rel = self.model.get_relation(PeerRelationName)
```

Comment:
please use `self.charm.peers_data....` - this will avoid having to do these checks etc...